### ISSUE TITLE
API idea: select!() macro for Selector literals

### ISSUE THREAD

**Issue by rvolgers** [causal-agent/scraper#53]
Inspired by [this blog post](https://blog.0xfa.be/building-a-backend-app-in-rust/) mentions some API struggles with writing a Rust backend program, and one of the mentioned things is that `Selector::parse('...').unwrap()` is not ergonomic when the selectors are constants and you are writing a lot of them.

How about a `select!()` macro to define a selector from a string, which removes the boilerplate for this case? There is precedent for this in the form of `vec![]` provided by the language and e.g. the `regex!()` macro provided by the regex crate.

**Comment by Boscop** [comment]
The first thing I did when using this crate is write a function that takes a &'static str, constructs a selector and unwraps the Result.
I would like if this function could be included in this crate. 
Why do you think it has to be a macro?

**Comment by teymour-aldridge** [comment]
> Why do you think it has to be a macro?

A macro is nicer in this case because it's easier to return compile errors which reduce the need for run-time debugging.

> The first thing I did when using this crate is write a function that takes a &'static str, constructs a selector and unwraps the Result.

It is probably useful in some cases to construct selectors dynamically (so either a function taking a `String` or `Cow<'static, str>` is helpful there – `&'_ str`'s are harder to manipulate because they are stack rather than heap references).

**Comment by TonalidadeHidrica** [comment]
FWIW I define the following macro that enables caching too.

```rust
use once_cell::sync::Lazy;
use scraper::Selector;

macro_rules! selector {
    ($e: expr) => {{
        static SELECTOR: Lazy<Selector> = Lazy::new(|| Selector::parse($e).unwrap());
        &*SELECTOR
    }};
}
```

**Comment by nathaniel-daniel** [comment]
> A macro is nicer in this case because it's easier to return compile errors which reduce the need for run-time debugging.

How would a macro return a compile-time error? 

> FWIW I define the following macro that enables caching too.

Wouldn't caching be the only benefit of a `selector!` style macro?

**Comment by teymour-aldridge** [comment]
> How would a macro return a compile-time error?

There's a compiler built-in macro called `[compile_error!](https://doc.rust-lang.org/stable/std/macro.compile_error.html)`; if a macro outputs that in the code it generates, the compiler fails the compilation and reports the error(s).

**Comment by nathaniel-daniel** [comment]
Sorry, I meant how would a macro know if a selector was invalid?

**Comment by TonalidadeHidrica** [comment]
Oops, sorry, I miswrote "too", but my prototype is the one which *only* enables caching.

**Comment by SirWindfield** [comment]
> ```rust
> use once_cell::sync::Lazy;
> use scraper::Selector;
> 
> macro_rules! selector {
>     ($e: expr) => {{
>         static SELECTOR: Lazy<Selector> = Lazy::new(|| Selector::parse($e).unwrap());
>         &*SELECTOR
>     }};
> }
> ```

I personally think that there isn't much value in adding this to the library (in its current form). It's easy enough that it can be added to the user's codebase instead.

As far as I can tell there is no way to actually check a static string at compile-time and throw a compilation-error if the string is not valid. Normal macros are just expanded, not evaluated. 

Maybe a proc-macro would do the trick (just the idea sketched out):

```rust
#[proc_macro]
pub fn make_answer(ts: TokenStream) -> TokenStream {
    if /* first and only argument is a static string */ {
		if Selector::parse(/* first argument */).is_ok() {
        	/* emit "Selector::parse(first arg).unwrap()" */
        } else {
			compile_error!("failed to compile selector string");
        }
    }
}
```

Probably not the cleanest solution, but I think it would work. It's easy enough to not even require complex library dependencies like `syn`, since the only check that has to be done is the "token is a static string" check.

**Comment by SirWindfield** [comment]
Here is a MVP: https://github.com/SirWindfield/scraper-proc-macro. Compilation fails for one of the tests as the proc-macro panics due to a wrong formatted selector string.

**Comment by cfvescovo** [comment]
> I personally think that there isn't much value in adding this to the library (in its current form). It's easy enough that it can be added to the user's codebase instead and is probably the place where something like this _should_ be added as well.

Agree. Maybe an external library, like the one @mainrs provided, is better
